packages/oid4vci-holder/src/agent/OID4VCIHolderService.ts

```diff
@@ -204,28 +232,329 @@ export const verifyCredential = async (args: IVerifyCredentialArgs, context: Req
+      }
+
+      // if we cannot find supported cryptographic suites, we just try with the first preference
+      const fallback = jwtCryptographicSuitePreferences[0]
+      console.log(`Warn: We could not determine the crypto suites from the server metadata, and will fallback to a default: ${fallback}`)
+      return fallback
+    }
+    // @ts-ignore
+    case 'ldp':
+    // @ts-ignore
+    case 'jwt_vc_json_ld':
+    case 'ldp_vc': {
+      const supportedPreferences: Array<string> = jsonldCryptographicSuitePreferences.filter((suite: string) => suites_supported.includes(suite))
+      if (supportedPreferences.length > 0) {
+        return supportedPreferences[0]
+      }
+
+      // if we cannot find supported cryptographic suites, we just try with the first preference
+      const fallback = jsonldCryptographicSuitePreferences[0]
+      console.log(`Warn: We could not determine the crypto suites from the server metadata, and will fallback to a default: ${fallback}`)
+      return fallback
+    }
+    default:
+      return Promise.reject(Error(`Credential format '${credentialSupported.format}' not supported`))
+  }
+}
+
+export const signatureAlgorithmFromKey = async (args: SignatureAlgorithmFromKeyArgs): Promise<SignatureAlgorithmEnum> => {
+  const { key } = args
+  return signatureAlgorithmFromKeyType({ type: key.type })
+}
+
+export const signJWT = async (args: SignJwtArgs): Promise<string> => {
+  const { identifier, header, payload, context, options } = args
+  const jwtOptions = {
+    ...options,
+    signer: await getSigner({ identifier, context }),
+  }
+
+  return createJWT(payload, jwtOptions, header)
+}
+
+export const getSigner = async (args: GetSignerArgs): Promise<Signer> => {
+  // TODO
+  const { identifier, context } = args
+  // TODO currently we assume an identifier only has one key
+  const key = identifier.keys[0]
+  // TODO See if this is mandatory for a correct JWT
+  const algorithm = await signatureAlgorithmFromKey({ key })
+
+  return async (data: string | Uint8Array): Promise<string> => {
+    const input = data instanceof Object.getPrototypeOf(Uint8Array) ? new TextDecoder().decode(data as Uint8Array) : (data as string)
+    return await context.agent.keyManagerSign({
+      keyRef: key.kid,
+      algorithm,
+      data: input,
+    })
+  }
+}
+
+export const signatureAlgorithmFromKeyType = (args: SignatureAlgorithmFromKeyTypeArgs): SignatureAlgorithmEnum => {
+  const { type } = args
+  switch (type) {
+    case 'Ed25519':
+    case 'X25519':
+      return SignatureAlgorithmEnum.EdDSA
+    case 'Secp256r1':
+      return SignatureAlgorithmEnum.ES256
+    case 'Secp256k1':
+      return SignatureAlgorithmEnum.ES256K
+    default:
+      throw new Error(`Key type '${type}' not supported`)
+  }
+}
+
+// TODO improve this conversion for jwt and jsonld, not a fan of current structure
+export const keyTypeFromCryptographicSuite = (args: KeyTypeFromCryptographicSuiteArgs): TKeyType => {
+  // TODO async?
```

Comment:
Async for a mapping function?

Reply:
no this is an old comment, should have been removed